### ISSUE TITLE
404 for https://civicdb.org/api/genes/${gene}?identifier_type=entrez_symbol when ${gene} is unknown

### ISSUE THREAD

**Issue by inodb** [griffithlab/civic-server#420]
Hey @susannasiebert ! Was hoping you could help us with an issue in cBioPortal pulling data from the civic API.

It's this issue:

https://github.com/cBioPortal/cbioportal/issues/3967

We can solve this on our end by 

(1) first checking which genes are supported by civic using the following query:

https://civicdb.org/api/genes/?identifier_type=entrez_symbol&count=9999999999999

Then only querying those genes in that list.

or on your end by:

(2) The civic endpoint `/genes/` should return an empty response only for the genes that fail. The other gene names should still work. 

Both could make sense. Since there are only 300 genes in CIVIC it might actually give a performance boost for us to only query those 300 genes as proposed in (1). It would however be less effort on our end if (2) were implemented. Let me know what you think.

**Comment by susannasiebert** [comment]
There are a couple of things going on here:
(1) We are currently uppercasing the search term so a search for `C11orf21` (which is in CIViC) will fail CIViC changes the search term to `C11ORF21`, which is not the official Entrez gene name.
(2) This search only succeeds for genes that have CIViC variants associated with them, so even if the gene `C11orf21` were successfully found, this would still return a 404 because there are no associated genes for it in CIViC. We (the CIViC team) need to discuss if we want to change this behavior.
(3) After talking to @acoffman we do think that this endpoint should return a 404 if someone searches by a gene name that doesn't exist, e.g. a nonsense terms like `foo`.

For some of the genes in your list, the search returns successful for me, e.g., `https://civicdb.org/api/genes/LSP1?identifier_type=entrez_symbol`. Not sure what is going on there.

**Comment by inodb** [comment]
@susannasiebert Thanks for the quick reply! I agree that it makes sense to return 404 when searching for one gene that is a nonsense term like `foo`. But when doing a batch query for multiple genes using the comma separated values syntax it might make more sense to only return empty/error response for those genes that are wrong?

**Comment by susannasiebert** [comment]
@inodb I apologize for the delayed response. After discussing this some more we decided we would change it so that a query with multiple search terms would only return the genes that were found in CIViC. This means that the list of genes might be shorter than the list of search terms if some of the search terms were not found. We would also continue to limit the results to only genes that have associated variants/evidence in CIViC. Will this solution work for you?

**Comment by inodb** [comment]
Thanks @susannasiebert ! That sounds like a good solution to me

**Comment by susannasiebert** [comment]
@inodb Digging more into the code, it looks like the suggested behavior is already implemented when searching for multiple terms, e.g., `https://civicdb.org/api/genes/ALK,BRAF,foo?identifier_type=entrez_symbol` works/doesn't return a 404. I only get a 404 when searching for single terms, e.g., `https://civicdb.org/api/genes/foo?identifier_type=entrez_symbol`. On your end, do you use the single term search or a multi-term search?

**Comment by inodb** [comment]
@susannasiebert ah ok - the behavior is indeed like that. Good point, thanks! Our issue was because of querying a gene with this `MIR-3676%2F3676` name. Thanks for resolving this!

**Comment by inodb** [comment]
@susannasiebert So actually this is prolly still good to fix. Right now when one of the gene's names in the multiterm search is this `MIR-3676%2F3676` (encoded version of `MIR-3676/3676` it returns 404 for the entire query. It seems to 404 when one of the gene names has the encoded `/` in the name

**Comment by susannasiebert** [comment]
Yeah, that is difficult because we'd need to be able to differentiate between forward slashes that are part of the URI and ones that are part of the gene name. I think we tried to support this previously but it broke other things so we had to revert the change. (Original issue: #361, "fixed" in commit e5010bd, reverted in PR #367)

**Comment by susannasiebert** [comment]
Another option would be to search by entrez id instead of entrez symbol, e.g., https://civicdb.org/api/genes/673,7157?identifier_type=entrez_id
